// devextreme/cjs/core/options/utils.js
/**
 * DevExtreme (cjs/core/options/utils.js)
 * Version: 24.2.7
 * Build date: Mon Apr 28 2025
 *
 * Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "convertRulesToOptions", {
    enumerable: true,
    get: function() {
        return _m_utils.convertRulesToOptions
    }
});
Object.defineProperty(exports, "createDefaultOptionRules", {
    enumerable: true,
    get: function() {
        return _m_utils.createDefaultOptionRules
    }
});
Object.defineProperty(exports, "deviceMatch", {
    enumerable: true,
    get: function() {
        return _m_utils.deviceMatch
    }
});
Object.defineProperty(exports, "getFieldName", {
    enumerable: true,
    get: function() {
        return _m_utils.getFieldName
    }
});
Object.defineProperty(exports, "getNestedOptionValue", {
    enumerable: true,
    get: function() {
        return _m_utils.getNestedOptionValue
    }
});
Object.defineProperty(exports, "getParentName", {
    enumerable: true,
    get: function() {
        return _m_utils.getParentName
    }
});
Object.defineProperty(exports, "normalizeOptions", {
    enumerable: true,
    get: function() {
        return _m_utils.normalizeOptions
    }
});
var _m_utils = require("../../__internal/core/options/m_utils");
