// devextreme-react/esm/pagination.js
/*!
 * devextreme-react
 * Version: 24.2.7
 * Build date: Mon Apr 28 2025
 *
 * Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-react
 */

"use client";
import * as React from "react";
import { memo, forwardRef, useImperativeHandle, useRef, useMemo } from "react";
import dxPagination from "devextreme/ui/pagination";
import { Component as BaseComponent } from "./core/component";
const Pagination = memo(forwardRef((props, ref) => {
    const baseRef = useRef(null);
    useImperativeHandle(ref, () => ({
        instance() {
            return baseRef.current?.getInstance();
        }
    }), [baseRef.current]);
    const subscribableOptions = useMemo(() => (["pageIndex", "pageSize"]), []);
    const independentEvents = useMemo(() => (["onContentReady", "onDisposing", "onInitialized"]), []);
    const defaults = useMemo(() => ({
        defaultPageIndex: "pageIndex",
        defaultPageSize: "pageSize",
    }), []);
    return (React.createElement((BaseComponent), {
        WidgetClass: dxPagination,
        ref: baseRef,
        subscribableOptions,
        independentEvents,
        defaults,
        ...props,
    }));
}));
export default Pagination;
export { Pagination };
