// devextreme-react/cjs/common/data.js
/*!
 * devextreme-react
 * Version: 24.2.7
 * Build date: Mon Apr 28 2025
 *
 * Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-react
 */

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setErrorHandler = exports.query = exports.ODataStore = exports.ODataContext = exports.LocalStore = exports.keyConverters = exports.isLoadResultObject = exports.isItemsArray = exports.isGroupItemsArray = exports.errorHandler = exports.EndpointSelector = exports.EdmLiteral = exports.DataSource = exports.CustomStore = exports.compileSetter = exports.compileGetter = exports.base64_encode = exports.ArrayStore = exports.applyChanges = void 0;
var data_1 = require("devextreme/common/data");
Object.defineProperty(exports, "applyChanges", { enumerable: true, get: function () { return data_1.applyChanges; } });
Object.defineProperty(exports, "ArrayStore", { enumerable: true, get: function () { return data_1.ArrayStore; } });
Object.defineProperty(exports, "base64_encode", { enumerable: true, get: function () { return data_1.base64_encode; } });
Object.defineProperty(exports, "compileGetter", { enumerable: true, get: function () { return data_1.compileGetter; } });
Object.defineProperty(exports, "compileSetter", { enumerable: true, get: function () { return data_1.compileSetter; } });
Object.defineProperty(exports, "CustomStore", { enumerable: true, get: function () { return data_1.CustomStore; } });
Object.defineProperty(exports, "DataSource", { enumerable: true, get: function () { return data_1.DataSource; } });
Object.defineProperty(exports, "EdmLiteral", { enumerable: true, get: function () { return data_1.EdmLiteral; } });
Object.defineProperty(exports, "EndpointSelector", { enumerable: true, get: function () { return data_1.EndpointSelector; } });
Object.defineProperty(exports, "errorHandler", { enumerable: true, get: function () { return data_1.errorHandler; } });
Object.defineProperty(exports, "isGroupItemsArray", { enumerable: true, get: function () { return data_1.isGroupItemsArray; } });
Object.defineProperty(exports, "isItemsArray", { enumerable: true, get: function () { return data_1.isItemsArray; } });
Object.defineProperty(exports, "isLoadResultObject", { enumerable: true, get: function () { return data_1.isLoadResultObject; } });
Object.defineProperty(exports, "keyConverters", { enumerable: true, get: function () { return data_1.keyConverters; } });
Object.defineProperty(exports, "LocalStore", { enumerable: true, get: function () { return data_1.LocalStore; } });
Object.defineProperty(exports, "ODataContext", { enumerable: true, get: function () { return data_1.ODataContext; } });
Object.defineProperty(exports, "ODataStore", { enumerable: true, get: function () { return data_1.ODataStore; } });
Object.defineProperty(exports, "query", { enumerable: true, get: function () { return data_1.query; } });
Object.defineProperty(exports, "setErrorHandler", { enumerable: true, get: function () { return data_1.setErrorHandler; } });
