// devextreme-react/cjs/common/index.js
/*!
 * devextreme-react
 * Version: 24.2.7
 * Build date: Mon Apr 28 2025
 *
 * Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-react
 */

"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Grids = exports.Export = exports.Data = exports.Core = exports.Charts = exports.setTemplateEngine = exports.Guid = exports.config = void 0;
var common_1 = require("devextreme/common");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return common_1.config; } });
Object.defineProperty(exports, "Guid", { enumerable: true, get: function () { return common_1.Guid; } });
Object.defineProperty(exports, "setTemplateEngine", { enumerable: true, get: function () { return common_1.setTemplateEngine; } });
exports.Charts = __importStar(require("./charts"));
exports.Core = __importStar(require("./core/index"));
exports.Data = __importStar(require("./data"));
exports.Export = __importStar(require("./export/index"));
exports.Grids = __importStar(require("./grids"));
