// devextreme-react/cjs/common/core/localization.js
/*!
 * devextreme-react
 * Version: 24.2.7
 * Build date: Mon Apr 28 2025
 *
 * Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-react
 */

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseNumber = exports.parseDate = exports.locale = exports.loadMessages = exports.formatNumber = exports.formatMessage = exports.formatDate = void 0;
var localization_1 = require("devextreme/common/core/localization");
Object.defineProperty(exports, "formatDate", { enumerable: true, get: function () { return localization_1.formatDate; } });
Object.defineProperty(exports, "formatMessage", { enumerable: true, get: function () { return localization_1.formatMessage; } });
Object.defineProperty(exports, "formatNumber", { enumerable: true, get: function () { return localization_1.formatNumber; } });
Object.defineProperty(exports, "loadMessages", { enumerable: true, get: function () { return localization_1.loadMessages; } });
Object.defineProperty(exports, "locale", { enumerable: true, get: function () { return localization_1.locale; } });
Object.defineProperty(exports, "parseDate", { enumerable: true, get: function () { return localization_1.parseDate; } });
Object.defineProperty(exports, "parseNumber", { enumerable: true, get: function () { return localization_1.parseNumber; } });
